# TP2: Machine Learning

In [1]:
import pandas as pd

# import ML packages
from sklearn.cluster import KMeans
from sklearn.datasets import make_blobs
from sklearn import linear_model
from sklearn import preprocessing

# import Plotting pckgs
from matplotlib import pyplot as plt
%matplotlib inline

In [2]:
trainingSet = pd.read_csv('../CSVs/improved_trip_train.csv')
testingSet = pd.read_csv('../CSVs/improved_trip_test.csv')
testingOriginal = pd.read_csv('../CSVs/trip_test.csv')
trainingSet.dtypes

id                     int64
duration               int64
start_date            object
start_dayOfWeek        int64
start_week             int64
start_quarter          int64
start_time            object
start_hour             int64
start_minute           int64
start_station_name    object
start_station_id       int64
end_date              object
end_dayOfWeek          int64
end_week               int64
end_quarter            int64
end_time              object
end_hour               int64
end_minute             int64
end_station_name      object
end_station_id         int64
bike_id                int64
subscription_type     object
zip_code              object
dtype: object

In [3]:
trainingSet.head()

id  duration  start_date  start_dayOfWeek  start_week  start_quarter  \
0  907649       396  2015-08-27                3          35              3   
1  384043       636  2014-07-28                0          31              3   
2  316176       334  2014-06-09                0          24              2   
3  618874       666  2015-01-26                0           5              1   
4  910977       318  2015-08-29                5          35              3   

  start_time  start_hour  start_minute  \
0   08:36:00           8            36   
1   22:06:00          22             6   
2   08:42:00           8            42   
3   16:55:00          16            55   
4   15:09:00          15             9   

                        start_station_name    ...     end_week end_quarter  \
0     Harry Bridges Plaza (Ferry Building)    ...           35           3   
1                           Market at 10th    ...           31           3   
2                        Market at Sansome    ...           24           2   
3  San Francisco Caltrain 2 (330 Townsend)    ...            5           1   
4                           Market at 10th    ...           35           3   

   end_time  end_hour  end_minute  \
0  08:43:00         8          43   
1  22:17:00        22          17   
2  08:47:00         8          47   
3  17:07:00        17           7   
4  15:14:00        15          14   

                                end_station_name  end_station_id  bike_id  \
0                         Embarcadero at Sansome              60      187   
1                           Washington at Kearny              46      417   
2                                  2nd at Folsom              62      281   
3  Temporary Transbay Terminal (Howard at Beale)              55      634   
4                             Powell Street BART              39      607   

  subscription_type  zip_code  
0        Subscriber     94602  
1        Subscriber     94133  
2        Subscriber     94107  
3        Subscriber     94602  
4        Subscriber     94709  

[5 rows x 23 columns]

In [4]:
trainingSet.shape

(549961, 23)

In [5]:
testingSet.head()

id  start_date  start_dayOfWeek  start_week  start_quarter start_time  \
0  504737  2014-10-18                5          42              4   11:25:00   
1  530846  2014-11-05                2          45              4   13:00:00   
2  813140  2015-06-18                3          25              2   17:34:00   
3  897674  2015-08-20                3          34              3   07:06:00   
4  322830  2014-06-13                4          24              2   08:46:00   

   start_hour  start_minute                        start_station_name  \
0          11            25                    Embarcadero at Sansome   
1          13             0                     Embarcadero at Folsom   
2          17            34  San Francisco Caltrain (Townsend at 4th)   
3           7             6         Civic Center BART (7th at Market)   
4           8            46   San Francisco Caltrain 2 (330 Townsend)   

   start_station_id    ...    end_week  end_quarter  end_time  end_hour  \
0                60    ...          42            4  12:12:00        12   
1                51    ...          45            4  13:09:00        13   
2                70    ...          25            2  17:37:00        17   
3                72    ...          34            3  07:15:00         7   
4                69    ...          24            2  08:57:00         8   

  end_minute               end_station_name  end_station_id bike_id  \
0         12  Powell at Post (Union Square)              71     426   
1          9      Broadway St at Battery St              82     454   
2         37                2nd at Townsend              61     370   
3         15                Townsend at 7th              65     451   
4         57          Embarcadero at Folsom              51     603   

   subscription_type  zip_code  
0           Customer     77009  
1         Subscriber     94132  
2         Subscriber     94107  
3         Subscriber     94582  
4         Subscriber     95014  

[5 rows x 22 columns]

In [6]:
trainingSet.duration.mean()

1113.4153549069842

### Let's study correlation between data

In [7]:
trainingSet.corr()['duration']

id                 -0.003133
duration            1.000000
start_dayOfWeek     0.018351
start_week          0.004002
start_quarter       0.003648
start_hour          0.002598
start_minute        0.002586
start_station_id   -0.007001
end_dayOfWeek       0.013753
end_week            0.002100
end_quarter         0.001376
end_hour            0.009626
end_minute         -0.000530
end_station_id     -0.006026
bike_id            -0.001569
Name: duration, dtype: float64

We can see that the duration is most correlated with:
    - start_dayOfWeek     0.018351
    - end_dayOfWeek       0.013753
    - end_hour            0.009626
    - start_station_id   -0.007001

In [8]:
columns = trainingSet.columns.tolist()
columns

['id',
 'duration',
 'start_date',
 'start_dayOfWeek',
 'start_week',
 'start_quarter',
 'start_time',
 'start_hour',
 'start_minute',
 'start_station_name',
 'start_station_id',
 'end_date',
 'end_dayOfWeek',
 'end_week',
 'end_quarter',
 'end_time',
 'end_hour',
 'end_minute',
 'end_station_name',
 'end_station_id',
 'bike_id',
 'subscription_type',
 'zip_code']

In [9]:
trainingSet.dtypes

id                     int64
duration               int64
start_date            object
start_dayOfWeek        int64
start_week             int64
start_quarter          int64
start_time            object
start_hour             int64
start_minute           int64
start_station_name    object
start_station_id       int64
end_date              object
end_dayOfWeek          int64
end_week               int64
end_quarter            int64
end_time              object
end_hour               int64
end_minute             int64
end_station_name      object
end_station_id         int64
bike_id                int64
subscription_type     object
zip_code              object
dtype: object

In [10]:
# Create linear regression object
regr = linear_model.LinearRegression()

In [11]:
trainingSet.shape

(549961, 23)

In [12]:
# Train the model using the training sets
regr.fit(trainingSet[['start_dayOfWeek', 'end_dayOfWeek', 'end_hour', 'start_station_id']], trainingSet.duration)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)

In [13]:
prediction = regr.predict(testingSet[['start_dayOfWeek', 'end_dayOfWeek', 'end_hour', 'start_station_id']])

In [14]:
prediction

array([ 1603.57291184,  1069.66886328,  1298.9958606 , ...,  1784.36773396,
        1544.50655641,   596.68919234])

In [15]:
duration = { 
    'id' : testingSet.id,
    'duration' : prediction}
df_duration = pd.DataFrame(duration, columns = ['id','duration'])

In [16]:
df_duration

id     duration
0       504737  1603.572912
1       530846  1069.668863
2       813140  1298.995861
3       897674   836.063931
4       322830  1126.487657
5       487841  1118.016993
6       677808  1150.710172
7       704449   835.598940
8       833587  1760.988055
9       420411  1534.923806
10      354777   868.292119
11       17809   883.850170
12      854667   783.069776
13      673325   311.008392
14      729900  1171.011148
15        4485  1342.114626
16      349394  1075.254624
17      784460  1342.114626
18      908431  1422.594044
19      805578  1660.769551
20      609571   356.374718
21      269385  1463.573842
22      227320   732.301734
23      658277  1353.944938
24      205479  1349.386058
25      429936  1503.354407
26      176953   839.888570
27       64226  1580.571078
28      489619  1062.225080
29      831189   986.769533
...        ...          ...
119968   60831  1106.186681
119969  875546   904.528992
119970  126578   133.757263
119971  244611  1055.699584
119972  110019  1044.247116
119973  114911  1033.712936
119974  649284  1115.597080
119975  719648   949.236528
119976  536952    27.294208
119977  579153  1105.905736
119978  619244   942.061996
119979  277802  1126.487657
119980  572524   634.406242
119981  273279   822.363902
119982  438379  1466.555645
119983  322903  1197.556676
119984  226880   869.415900
119985   34690  1087.646827
119986  581506  1412.168458
119987  365644   660.951770
119988   18373  1257.454172
119989  646921   637.949935
119990  835970   512.947026
119991  100540   862.425413
119992  503083  1126.487657
119993  274799   714.323770
119994  464786   738.827230
119995  172825  1784.367734
119996  641204  1544.506556
119997  893371   596.689192

[119998 rows x 2 columns]

# HISTORICO

In [87]:
#Armamos un Data Frame mas comodo para trabajar. Nuestro objetivo es calcular el HISTORICO de cada viaje
#Historico: promedio de duracion de viajes entre una estacion y otra.
trainingShort = trainingSet.loc[:,['id','duration','start_station_name','start_station_id','end_station_name','end_station_id']]
trainingShort

id  duration                             start_station_name  \
0       907649       396           Harry Bridges Plaza (Ferry Building)   
1       384043       636                                 Market at 10th   
2       316176       334                              Market at Sansome   
3       618874       666        San Francisco Caltrain 2 (330 Townsend)   
4       910977       318                                 Market at 10th   
5       522083       337                                Townsend at 7th   
6       880809       394                              2nd at South Park   
7       488938       766                  Powell at Post (Union Square)   
8       899522       531                                  2nd at Folsom   
9       737380       267                                  Market at 4th   
10      291053       766       San Francisco Caltrain (Townsend at 4th)   
11      332574       416                              Market at Sansome   
12      432022      4120           Harry Bridges Plaza (Ferry Building)   
13      398366       525        San Francisco Caltrain 2 (330 Townsend)   
14      863630       301                                2nd at Townsend   
15      644854       311                             Powell Street BART   
16      390681       528                                Beale at Market   
17      224386      4675           Harry Bridges Plaza (Ferry Building)   
18      246380       203                          Embarcadero at Bryant   
19      350371       916  Temporary Transbay Terminal (Howard at Beale)   
20      429032       246                              2nd at South Park   
21      479700       431                              Market at Sansome   
22      529387       722                                  Howard at 2nd   
23      777536       275  Temporary Transbay Terminal (Howard at Beale)   
24      769331       560                              Steuart at Market   
25      537257       697          Rengstorff Avenue / California Street   
26      298292      1595           Harry Bridges Plaza (Ferry Building)   
27      485095       470                                2nd at Townsend   
28      109338       372              San Jose Diridon Caltrain Station   
29      416274       472                Grant Avenue at Columbus Avenue   
...        ...       ...                                            ...   
549931   10841      1111                              Market at Sansome   
549932  155867       523                              2nd at South Park   
549933  322105       906                         Embarcadero at Sansome   
549934  865329       600           Harry Bridges Plaza (Ferry Building)   
549935  419180       290                      Broadway St at Battery St   
549936  288255       255        San Francisco Caltrain 2 (330 Townsend)   
549937   27175       933                          Embarcadero at Bryant   
549938  716966       180                                 Market at 10th   
549939  422409       478              Civic Center BART (7th at Market)   
549940  599416       682                         Embarcadero at Sansome   
549941  860647       878                         SJSU 4th at San Carlos   
549942  174919       523                                 Market at 10th   
549943  449873       266                 Mountain View Caltrain Station   
549944  711539       771  Temporary Transbay Terminal (Howard at Beale)   
549945  150716       370                                 Market at 10th   
549946  172164       462           Harry Bridges Plaza (Ferry Building)   
549947  444391       830                              Market at Sansome   
549948   53552       976                          Washington at Kearney   
549949  673387       659        San Francisco Caltrain 2 (330 Townsend)   
549950  102255       473                              Market at Sansome   
549951  629989      1098  Temporary Transbay Terminal (Howard at Beale)   
549952  565782       394             

## Modo prueba para probar el algoritmo

In [40]:
prueba = {
    'id': [1,2,3,4,5],
    'duration':[31,21,10,15,6],
    'start':[1,1,3,2,3],
    'end':[2,3,2,1,2]
}

dfPrueba = pd.DataFrame(prueba,columns = ['id','duration','start','end'])
dfPrueba

id  duration  start  end
0   1        31      1    2
1   2        21      1    3
2   3        10      3    2
3   4        15      2    1
4   5         6      3    2

In [56]:
#Armamos unas listas cuyos elementos son las ID de las estaciones de llegada y salida
#Eliminamos los repetidos
listaStart = []
listaEnd = []
for i in list(dfPrueba.start.values):
    if i not in listaStart:
        listaStart.append(i)
for i in list(dfPrueba.end.values):
    if i not in listaEnd:
        listaEnd.append(i)

In [73]:
#Creamos una Tupla de (idSalida,idLlegada,historico)
#Armamos una lista con las tuplas.
#No agregamos los Nan. En este caso seran las tuplas donde no existe un historico, debido a que entre esas dos estaciones 
#nunca se llevo a cabo un viaje
listaHistorico = []
for i in listaStart:
    for j in listaEnd:
        df = dfPrueba[(dfPrueba['start'] == i) & (dfPrueba['end'] == j)]
        historico = df.duration.mean()
        if (not(math.isnan(historico))):
            listaHistorico.append([i,j,historico])
        
listaHistorico

[[1, 2, 31.0], [1, 3, 21.0], [3, 2, 8.0], [2, 1, 15.0]]

In [77]:
starStationId = []
endStationId = []
historical = []
for x in listaHistorico:
    starStationId.append(x[0])
    endStationId.append(x[1])
    historical.append(x[2])

data = {
    'start' : starStationId,
    'end' : endStationId,
    'historical' : historical,
}

dfData = pd.DataFrame(data,columns = ['start','end','historical'])
dfData

start  end  historical
0      1    2          31
1      1    3          21
2      3    2           8
3      2    1          15

In [79]:
result = pd.merge(dfPrueba,dfData,on =['start','end'],how = 'inner')
result

id  duration  start  end  historical
0   1        31      1    2          31
1   2        21      1    3          21
2   3        10      3    2           8
3   5         6      3    2           8
4   4        15      2    1          15

## Ahora la prueba de fuego con el Data Frame del Training Reducido

In [90]:
listaStart = []
listaEnd = []
for i in list(trainingShort.start_station_id.values):
    if i not in listaStart:
        listaStart.append(i)
for i in list(trainingShort.end_station_id.values):
    if i not in listaEnd:
        listaEnd.append(i)


In [91]:
#Tarda un rato.
listaHistorico = []
for i in listaStart:
    for j in listaEnd:
        df = trainingShort[(trainingShort['start_station_id'] == i) & (trainingShort['end_station_id'] == j)]
        historico = df.duration.mean()
        if (not(math.isnan(historico))):
            listaHistorico.append([i,j,historico])
        
listaHistorico

[[50, 60, 1254.993877147936],
 [50, 46, 2679.90099009901],
 [50, 62, 2177.8607350096713],
 [50, 55, 502.29606625258799],
 [50, 39, 1938.610909090909],
 [50, 70, 903.78094462540719],
 [50, 61, 662.55546241967033],
 [50, 75, 955.34452296819791],
 [50, 67, 2432.1732026143791],
 [50, 72, 7406.0220588235297],
 [50, 51, 1104.3922651933701],
 [50, 76, 1278.5260663507108],
 [50, 50, 6343.4900849858359],
 [50, 66, 3062.5027027027027],
 [50, 54, 788.76531671858777],
 [50, 64, 842.403162055336],
 [50, 65, 1070.6219047619047],
 [50, 45, 846.55012224938878],
 [50, 77, 834.04506699147385],
 [50, 82, 1168.3443396226414],
 [50, 74, 4673.119565217391],
 [50, 68, 968.3078101071975],
 [50, 71, 2404.4898477157362],
 [50, 56, 2053.1680000000001],
 [50, 49, 453.73065015479875],
 [50, 57, 925.07014028056108],
 [50, 63, 661.69466882067854],
 [50, 48, 1801.127619047619],
 [50, 69, 1076.7346760070052],
 [50, 47, 1400.2985074626865],
 [50, 42, 1246.1496062992126],
 [50, 41, 1677.0087336244542],
 [50, 73, 2730.59

In [92]:
starStationId = []
endStationId = []
historical = []
for x in listaHistorico:
    starStationId.append(x[0])
    endStationId.append(x[1])
    historical.append(x[2])

data = {
    'start_station_id' : starStationId,
    'end_station_id' : endStationId,
    'historical' : historical,
}

dfData = pd.DataFrame(data,columns = ['start_station_id','end_station_id','historical'])
dfData

start_station_id  end_station_id    historical
0                   50              60   1254.993877
1                   50              46   2679.900990
2                   50              62   2177.860735
3                   50              55    502.296066
4                   50              39   1938.610909
5                   50              70    903.780945
6                   50              61    662.555462
7                   50              75    955.344523
8                   50              67   2432.173203
9                   50              72   7406.022059
10                  50              51   1104.392265
11                  50              76   1278.526066
12                  50              50   6343.490085
13                  50              66   3062.502703
14                  50              54    788.765317
15                  50              64    842.403162
16                  50              65   1070.621905
17                  50              45    846.550122
18                  50              77    834.045067
19                  50              82   1168.344340
20                  50              74   4673.119565
21                  50              68    968.307810
22                  50              71   2404.489848
23                  50              56   2053.168000
24                  50              49    453.730650
25                  50              57    925.070140
26                  50              63    661.694669
27                  50              48   1801.127619
28                  50              69   1076.734676
29                  50              47   1400.298507
...                ...             ...           ...
1713                38              35   2153.254902
1714                38              30   4517.000000
1715                38              38  22870.126582
1716                21              34   3440.000000
1717                21              37   3998.500000
1718                21              25   1002.133333
1719                21              22    303.717647
1720                21              35   3429.571429
1721                21              21   4655.500000
1722                21              23    729.000000
1723                21              83    972.500000
1724                21              26  10441.000000
1725                21              24   2521.285714
1726                83              34   2517.500000
1727                83              25    485.000000
1728                83              22    601.044444
1729                83              23    505.111111
1730                83              83   8959.275862
1731                83              26   1947.000000
1732                83              24    708.909091
1733                23              37  12731.000000
1734                23              25    505.884615
1735                23              22   2730.981818
1736                23              35   3206.000000
1737                23              21    751.166667
1738                23              23   6799.940299
1739                23              83   1755.375000
1740                23              26   1089.866667
1741                23              24   1013.030303
1742                23              38   1860.000000

[1743 rows x 3 columns]

In [94]:
trainingShort = pd.merge(trainingShort,dfData,on =['start_station_id','end_station_id'],how = 'inner')
trainingShort

id  duration                       start_station_name  \
0       907649       396     Harry Bridges Plaza (Ferry Building)   
1       807007       421     Harry Bridges Plaza (Ferry Building)   
2       343864      9624     Harry Bridges Plaza (Ferry Building)   
3       451164      1670     Harry Bridges Plaza (Ferry Building)   
4       101973       484     Harry Bridges Plaza (Ferry Building)   
5       502639       363     Harry Bridges Plaza (Ferry Building)   
6       842301       446     Harry Bridges Plaza (Ferry Building)   
7       300655       385     Harry Bridges Plaza (Ferry Building)   
8       781372       388     Harry Bridges Plaza (Ferry Building)   
9       727968       517     Harry Bridges Plaza (Ferry Building)   
10      589703       336     Harry Bridges Plaza (Ferry Building)   
11      772641       297     Harry Bridges Plaza (Ferry Building)   
12      341481      2004     Harry Bridges Plaza (Ferry Building)   
13      761200       638     Harry Bridges Plaza (Ferry Building)   
14      691315       841     Harry Bridges Plaza (Ferry Building)   
15      643647       434     Harry Bridges Plaza (Ferry Building)   
16      336619       468     Harry Bridges Plaza (Ferry Building)   
17      626169       391     Harry Bridges Plaza (Ferry Building)   
18       21102      1556     Harry Bridges Plaza (Ferry Building)   
19        9904      1168     Harry Bridges Plaza (Ferry Building)   
20      669002       441     Harry Bridges Plaza (Ferry Building)   
21      824097       572     Harry Bridges Plaza (Ferry Building)   
22      826895      3218     Harry Bridges Plaza (Ferry Building)   
23      582986       537     Harry Bridges Plaza (Ferry Building)   
24      624338       669     Harry Bridges Plaza (Ferry Building)   
25      281755       373     Harry Bridges Plaza (Ferry Building)   
26      645780       491     Harry Bridges Plaza (Ferry Building)   
27      628137       363     Harry Bridges Plaza (Ferry Building)   
28      192809     14530     Harry Bridges Plaza (Ferry Building)   
29      390753       388     Harry Bridges Plaza (Ferry Building)   
...        ...       ...                                      ...   
549931  631148      2872                 Stanford in Redwood City   
549932  359747      3347                         Broadway at Main   
549933   42493       236            Redwood City Caltrain Station   
549934  344146     72535               Palo Alto Caltrain Station   
549935   59290     12539           Mountain View Caltrain Station   
549936  656533      1711    Rengstorff Avenue / California Street   
549937  851553      1304    Rengstorff Avenue / California Street   
549938  780017      1317    Rengstorff Avenue / California Street   
549939  813868       782    Rengstorff Avenue / California Street   
549940  305297      6848                      San Salvador at 1st   
549941  305298      6853                      San Salvador at 1st   
549942  305300      6799                      San Salvador at 1st   
549943  385606      3553          California Ave Caltrain Station   
549944  208887     75844        San Jose Diridon Caltrain Station   
549945  370919     19216            Redwood City Caltrain Station   
549946  575830      3759            Redwood City Caltrain Station   
549947  421121     54178                  Mountain View City Hall   
549948   87473      3836                  Mountain View City Hall   
549949  402004      4333                   University and Emerson   
549950  848480      4150                   University and Emerson   
549951  383866      4072          California Ave Caltrain Station   
549952   17718     19076  San Francisco Caltrain 2 (330 Townsend)   
549953  812706     12731                  San Mateo County Center   
549954  254856     22361            Redwood City Caltrain Station   
549955  801195      1983              Redwood City Public Library   
549956  198755      1463              Redwood City Public Library   
549957  17